### PR TITLE
chore: serve specify dependency services

### PR DESCRIPTION
**Pull request by bojiang**
For yatai:
```
> bentoml start-http-server . --service-name Preprocessing --port 3000
> bentoml start-http-server . --service-name iris_classifier --port 4000 --depends Preprocessing=http://localhost:3000
```

### CODE REVIEW

File: src/bentoml_cli/start.py [hunk 1]
```diff
@@ -173,27 +169,49 @@ def start_http_server(  # type: ignore (unused warning)
...
+                )
+            else:
+                from bentoml.start import start_runner_server
+
+                if bind is not None:
+                    parsed = urlparse(bind)
+                    assert parsed.scheme == "tcp"
+                    host = parsed.hostname or host
+                    port = parsed.port or port
+
```
Comment (bojiang) [inline src/bentoml_cli/start.py:203]:
```suggestion
```

File: src/_bentoml_sdk/service/factory.py [hunk 2]
```diff
@@ -112,26 +116,25 @@ def _get_caller_module(self) -> str:
...
+        if attr_name not in self.dependencies:
+            if attr_name in self.all_services():
+                return self.all_services()[attr_name]
+            else:
+                raise ValueError(f"Service {attr_name} not found")
+        if path:
+            return self.dependencies[attr_name].on.find_dependent(path)
+        return self.dependencies[attr_name].on
+
+    @lru_cache(maxsize=1)
```
Comment (frostming) [inline src/_bentoml_sdk/service/factory.py:132]: Prefer `cached_property`

File: src/_bentoml_sdk/service/factory.py [hunk 3]
```diff
@@ -112,26 +116,25 @@ def _get_caller_module(self) -> str:
...
-            _import_str = self.import_string
-        _collected.add(self.name)
-        for name, dependency in self.dependencies.items():
-            services.extend(
-                dependency.on.all_services(
-                    _collected=_collected, _import_str=f"{_import_str}.{name}"
-                )
-            )
+    @lru_cache
+    def find_dependent(self, name_or_path: str) -> Service[t.Any]:
```
Comment (frostming) [inline src/_bentoml_sdk/service/factory.py:120]:
```suggestion
    def find_dependency(self, name_or_path: str) -> Service[t.Any]:
```

File: src/_bentoml_sdk/service/factory.py [hunk 4]
```diff
@@ -233,7 +236,8 @@ def inject_config(self) -> None:
         # XXX: ensure at least one item to make `flatten_dict` work
         override_defaults = {
             "services": {
-                svc.name: (svc.config or {"workers": 1}) for svc in self.all_services()
+                svc.name: (svc.config or {"workers": 1})
+                for svc in self.all_services().values()
```
Comment (frostming) [inline src/_bentoml_sdk/service/factory.py:240]:
```suggestion
                name: (svc.config or {"workers": 1})
                for name, svc in self.all_services().items()
```

File: src/_bentoml_impl/server/serving.py [hunk 5]
```diff
@@ -218,24 +197,30 @@ def serve_http(
...
-        if not standalone and not development_mode:
-            for name, dep in svc.dependencies.items():
-                dep_key = dep.cache_key()
-                if dep_key in dependency_map:
+
+        if service_name:
+            svc = svc.find_dependent(service_name)
+        elif development_mode:
+            pass
+        else:
```
Comment (frostming) [inline src/_bentoml_impl/server/serving.py:205]:
```suggestion
        elif not development_mode:
```

File: src/_bentoml_impl/server/serving.py [hunk 6]
```diff
@@ -218,24 +197,30 @@ def serve_http(
...
+            for dep_svc in svc.all_services().values():
+                if dep_svc.name == svc.name:
+                    continue
+                if dep_svc.name in dependency_map:
```
Comment (frostming) [inline src/_bentoml_impl/server/serving.py:209]:
```suggestion
            for name, dep_svc in svc.all_services().items():
                if name == svc.name:
                    continue
                if name in dependency_map:
```

File: src/_bentoml_impl/server/serving.py [hunk 7]
```diff
@@ -218,24 +197,30 @@ def serve_http(
...
                     allocator,
-                    f"{svc.import_string}.{name}",
                     str(bento_path.absolute()),
                 )
-                watchers.extend(new_watchers)
-                sockets.extend(new_sockets)
-                dependency_map[dep_key] = uri
+                watchers.append(new_watcher)
+                sockets.append(new_socket)
+                dependency_map[dep_svc.name] = uri
```
Comment (frostming) [inline src/_bentoml_impl/server/serving.py:223]:
```suggestion
                dependency_map[name] = uri
```

File: src/_bentoml_sdk/service/factory.py (hunk 2 above)
Comment (bojiang) [reply src/_bentoml_sdk/service/factory.py:132]: Too lazy to modify all references

File: src/_bentoml_sdk/service/factory.py (hunk 3 above)
Comment (bojiang) [reply src/_bentoml_sdk/service/factory.py:121]: dependency is the relation.
dependent is the target itself. Service is a dependent, not a relation

**Comment by bojiang** [comment]
@frostming lgty?

**Review by frostming — APPROVED** [review]

**Pull request merged**